src/Workspaces/Core/Portable/CodeFixesAndRefactorings/DocumentBasedFixAllProviderHelpers.cs

```diff
@@ -40,32 +42,52 @@ internal static class DocumentBasedFixAllProviderHelpers
-        }
+            // First, iterate over all contexts, and collect all the changes for each of them.  We'll be making a lot of
+            // calls to the remote server to compute diagnostics and changes.  So keep a single connection alive to it
+            // so we never resync or recompute anything.
+            using var _2 = await RemoteKeepAliveSession.CreateAsync(solution, originalFixAllContext.CancellationToken).ConfigureAwait(false);
 
-        return currentSolution;
-    }
+            foreach (var fixAllContext in fixAllContexts)
+            {
+                Contract.ThrowIfFalse(
+                    fixAllContext.Scope is FixAllScope.Document or FixAllScope.Project or FixAllScope.ContainingMember or FixAllScope.ContainingType);
 
-    private static async Task<Solution> FixSingleContextAsync<TFixAllContext>(
-        Solution currentSolution,
-        TFixAllContext fixAllContext,
-        IProgress<CodeAnalysisProgress> progressTracker,
-        Func<TFixAllContext, IProgress<CodeAnalysisProgress>, Task<Dictionary<DocumentId, (SyntaxNode? node, SourceText? text)>>> getFixedDocumentsAsync)
-        where TFixAllContext : IFixAllContext
-    {
-        // First, compute and apply the fixes.
-        var docIdToNewRootOrText = await getFixedDocumentsAsync(fixAllContext, progressTracker).ConfigureAwait(false);
+                // TODO: consider computing this in parallel.
+                var singleContextDocIdToNewRootOrText = await getFixedDocumentsAsync(fixAllContext, progressTracker).ConfigureAwait(false);
+                allContextsDocIdToNewRootOrText.AddRange(singleContextDocIdToNewRootOrText);
+            }
+        }
 
-        // Then, cleanup the new doc roots, and apply the results to the solution.
-        currentSolution = await CleanupAndApplyChangesAsync(progressTracker, currentSolution, docIdToNewRootOrText, fixAllContext.CancellationToken).ConfigureAwait(false);
+        // Next, go and insert those all into the solution so all the docs in this particular project point at
+        // the new trees (or text).  At this point though, the trees have not been cleaned up.  We don't cleanup
+        // the documents as they are created, or one at a time as we add them, as that would cause us to run
+        // cleanup on N different solution forks (which would be very expensive).  Instead, by adding all the
+        // changed documents to one solution, and then cleaning *those* we only perform cleanup semantics on one
+        // forked solution.
+        var currentSolution = solution;
+        foreach (var (docId, (newRoot, newText)) in allContextsDocIdToNewRootOrText)
+        {
+            currentSolution = newRoot != null
+                ? currentSolution.WithDocumentSyntaxRoot(docId, newRoot)
+                : currentSolution.WithDocumentText(docId, newText!);
+        }
```

Comment:
we then go and form the next full solution, with all those initial changes applied.